### PR TITLE
fix: introduce modern typings for TS >4

### LINKED ISSUE #378: bug(core): Typescript error using animationName inside of a nested selector
repro: https://codesandbox.io/s/elegant-moon-odwjen?file=/src/App.tsx

When using `animationName` inside a nested selector, there are unexpected typescript errors:

```
(property) "&[data-foo]": {
    animationName: {
        from: {
            opacity: number;
        };
        to: {
            opacity: number;
        };
    }[];
}
Type '{ animationName: { from: { opacity: number; }; to: { opacity: number; }; }[]; }' is not assignable to type 'string | number | (string | number)[] | GriffelStylesCSSObjectCustomL2'.
  Type '{ animationName: { from: { opacity: number; }; to: { opacity: number; }; }[]; }' is not assignable to type 'string'.ts(2322)
```

### PR DESCRIPTION
**Pull request by layershifter**
Fixes #378.

This PR introduces modern typings for TypeScript 4.1 and higher as this version supports recursive types references. This allow to simplify typings and remove weird workarounds. Also newer typings make TS errors more precise (check changes in the fixtures) 👍 

To keep compatibility with older versions of TS older typings are kept as legacy. To resolve them `typesVersions` field is used.

### CODE REVIEW

File: e2e/typescript/src/assets/fixture.ts [hunk 1]
```diff
@@ -103,20 +109,30 @@ assertType({
...
+  overflow: '1',
+});
+assertType({
+  // @ts-expect-error "paddingLeft" cannot be numeric value
+  paddingLeft: 5,
+});
+assertType({
+  // @ts-expect-error "0" is invalid value for "color"
+  color: 0,
+});
```
Comment (layershifter) [inline e2e/typescript/src/assets/fixture.ts:135]: Just formatting change to detect errors on an exact line.

File: packages/core/project.json [hunk 2]
```diff
@@ -31,6 +31,14 @@
...
       }
     },
+    "build": {
+      "executor": "@nrwl/workspace:run-commands",
+      "dependsOn": [{ "target": "prebuild", "projects": "self" }],
+      "options": {
+        "commands": [{ "command": "node tools/createLegacyTypesForCore.mjs" }],
+        "outputPath": ["dist/packages/core/types-legacy"]
+      }
+    },
```
Comment (layershifter) [inline packages/core/project.json:41]: Runs the new script that copies all `.d.ts` files to `types-legacy` and then updates imports.

File: e2e/utils/src/createTempDir.ts [hunk 3]
```diff
@@ -3,7 +3,7 @@ import * as logSymbols from 'log-symbols';
 
 export function createTempDir(prefix: string) {
   // "Unsafe" means delete even if it still has files inside (our desired behavior)
-  const tempDir = tmp.dirSync({ prefix: 'nextjs', unsafeCleanup: true }).name;
+  const tempDir = tmp.dirSync({ prefix, unsafeCleanup: true }).name;
```
Comment (layershifter) [inline e2e/utils/src/createTempDir.ts:6]: Oops

File: packages/style-types/src/legacy/index.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+export type { GriffelAnimation, GriffelStyle } from './makeStyles';
```
Comment (layershifter) [inline packages/style-types/src/legacy/index.ts:1]: Copied from previous typings.

**Review by ling1726 — APPROVED** [review]

**Pull request merged**